csharp/ql/src/experimental/Security Features/campaign/Solorigate-Readme.md

```diff
@@ -0,0 +1,121 @@
+
+It can be evaded by including any statements in the catch block, but the existing cs/catch-of-all-exceptions query can be used to catch that evasion.  That query’s findings will be a superset of this query, as it will catch many legitimate situations where the exception was anticipated and the catch block can take action that would make swallowing the exception and continuing on relatively risk free.  By looking at empty catch statements, this query looks for the exact pattern in the Solorigate implant and will generate fewer findings to review.
+
+### cs/backdoor/dangerous-native-functions
+
+**Implemented in:** DangerousNativeFunctionCalls.ql
+
+This query looks for several native Windows API that might be present in backdoor functionality, including several used in Solorigate.  It is also more broadly useful in looking for a technique that can be used to evade queries looking for malicious execution of processes by calling the native platform libraries instead of the .Net Class libraries.  We are including it to act as inspiration for other query developers to think about how they can query not just for backdoor patterns but also how they can hunt for attempts to then evade those queries.
+
+While the results of this query are useful on their own, both to cast a wide net to find this evasion technique, and to review why the developers chose the more error prone PInvoke implementation in benign findings, this query was kept simple specifically to use as an example. It can easily be combined with Data Flow or Taint Flow by adding a call to the isExternMethod predicate in DangerousNativeFunctionCalls to an overridden isSink implementation.  With this relatively simple addition, flow from untrusted sources like web requests into the potentially dangerous API can easily be detected, and at the very least represent a potential security vulnerability that should be reviewed, but may also represent a C2 implementation attempting to evade searches for the standard Class Library functions.
+
+## Build CodeQL Databases from the production Build Servers/Pipelines
+
+The malicious modifications of SolarWinds’ code took place on the build server (more details from [CrowdStrike](https://www.crowdstrike.com/blog/sunspot-malware-technical-analysis/)), so CodeQL databases used for analysis should ideally be built from the same build servers.  CodeQL utilizes the same Roslyn Compiler as MSBuild, so absent the malicious actor specifically building checks for the presence of CodeQL, it is likely the injection technique utilized by the malicious actor would replicate when CodeQL was run.  Alternatively, if building CodeQL databases in an independent environment, other techniques can be used to validate that the code that was compiled into the final binaries is the same as the source code in the source repository.
+
+> ### FYI
```

Comment:
Are the quotes (using `>`) intentional here?

Reply:
yes. Would you recommend me to remove it?

Reply:
BTW. JBW gave a new version of the doc.